feat: add intelligent tab grouping by similarity

matteomekhail opened the pull request:
This feature automatically organizes workspace tabs into categories based on their content and domain.

How it works:
- Analyzes all unpinned tabs in the active workspace
- Uses keyword matching to categorize tabs by topic (Development, Social, Shopping, Video, News, etc.)
- Groups similar tabs together with visual category headers (Similar to Arc Browser)

The grouping is triggered via a new 'Group' button in the workspace toolbar.

Is not probably the best way to categorize it, a better future way would be to use AI to generate the categories based on the pages, instead of having a map with keyword, but this is definetely a cheaper way.

Localization support added for: en-US, en-GB, it (AI generated Language files, IT as is my main language)

matteomekhail committed:
feat: add intelligent tab grouping by similarity

This feature automatically organizes workspace tabs into categories based on their content and domain.

How it works:
- Analyzes all unpinned tabs in the active workspace
- Uses intelligent keyword matching to categorize tabs by topic (Development, Social, Shopping, Video, News, etc.)
- Groups similar tabs together with visual category headers
- Maintains logical ordering with priority-based sorting
- New tabs automatically stay above categorized groups

The grouping is triggered via a new 'Group' button in the workspace toolbar.
Categories include: Development, Social Media, Shopping, Video/Streaming, News, Productivity, Finance, Gaming, and more.

Localization support added for: en-US, en-GB, it

matteomekhail commented:
<img width="285" height="1017" alt="image" src="https://github.com/user-attachments/assets/5464e8a0-6708-46cc-80c6-d3cf8f8c6461" />

This is how the "Group" button is implemented in the UI, and also how the categories are shown in the sidebar.

mr-cheffy commented:
Have you tested this by the way?

matteomekhail commented:
> Have you tested this by the way?

Yes i did tested, but my bad, I've let AI do some things and didn't realized it forced pushed, I fixed it by restoring the working version, i'm really sorry for the confusion.

Here is a video showing the functionality:

https://github.com/user-attachments/assets/c83f04a2-c331-476c-b988-ec0192c4fd7a

mr-cheffy commented:
I think we should be using Firefox's local llm for this, no?

ZiedYousfi commented:
This is noisier than it needs to be. 

Also the list feels way too subjective. The Internet is way too big to be condensed and grouped in that way.

It also means that if we want to change the list we have to release a new version.

matteomekhail commented:
> I think we should be using Firefox's local llm for this, no?

didn't know i could do that, will implement that now

mr-cheffy commented:
> This is noisier than it needs to be.
> 
> Also the list feels way too subjective. The Internet is way too big to be condensed and grouped in that way.
> 
> It also means that if we want to change the list we have to release a new version.

Yep, I believe having a small, local LLM that just takes a look at tab titles and domains could serve just as good. Good thing about Firefox's LLM is that its not intrusive, its private and doesn't really do anything unless it being explicitly called. IIRC they now use the GPU for them as well? Meaning they are faster and less resource intensive.

@matteomekhail read: https://firefox-source-docs.mozilla.org/toolkit/components/ml/api.html

mr-cheffy committed:
Merge branch 'dev' into feature/group-tabs

matteomekhail commented:
Been trying to implement this LLM model solution in the last couple of days, and none of those can give me good result

BGE-base (or even small): 
Pros: stronger embeddings, fewer random merges.
Cons: still splits similar brands (Audi/Mercedes), many micro-clusters 

all-MiniLM-L6-v2:
Pros: model tailored for clustering, lightweight.
Cons: extremely sensitive threshold; based on configuration, either single-tab clusters, otherwise mixes distant domains

distilgpt2 :
Pros: basically none lol, just very light and fast
Cons: not instruction-tuned → needs heavy prompt engineering/parsing; still overuses the “Other” category, basically it can't understand how to categorize them

All the other Xenova model don't do the job as well, either they not able to generate the right categories title, or they will not categorize correctly ( ad example not merging adidas, nike, gucci or Lv under the same category)

The only approach that i've tried and it sorta work is a domain based, where i just look at the URL and merge them if they have the same based (merge all the yt tabs ad example) but that will not create a "video" category including other video provider like vimeo (as is a different domain)

I really want to implement this feature in the best possible way as it was my favorite feature on Arc, so if anyone has any idea or model to try i'll be very happy to

By "cluster" i mean a collection of tabs under a URL
Thanks for making the PR a draft! i should have done it sooner @mr-cheffy

12th-devs commented:
@matteomekhail I'm not sure if it could be any help, but I vibecoded a project like this that used two of the models Firefox uses to sort tabs (as I had the same issues as you with the other model) and had amazing results. I learned quite a lot about how it all worked in the process, so if I could help in any way, please feel free to reach out.

https://github.com/Vertex-Mods/Zen-Tidy-Tabs

12th-devs commented:
I would also look on searchfox and try to find the prompt they use for naming the tabs, as I don't remember off the top of my head

matteomekhail commented:
> @matteomekhail I'm not sure if it could be any help, but I vibecoded a project like this that used two of the models Firefox uses to sort tabs (as I had the same issues as you with the other model) and had amazing results. I learned quite a lot about how it all worked in the process, so if I could help in any way, please feel free to reach out.
> 
> https://github.com/Vertex-Mods/Zen-Tidy-Tabs



wow i just saw the screenshot and it looks good! i'll have a look at the code tomorrow, thank you!

mr-cheffy commented:
Firefox does have an auto-grouping feature, might be worth to checkout how they do it:

https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs

12th-devs commented:
> Firefox does have an auto-grouping feature, might be worth to checkout how they do it:
> 
> https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs

~~That's the model I use in the script~~

Ignore this, two different things.

12th-devs commented:
> > @matteomekhail I'm not sure if it could be any help, but I vibecoded a project like this that used two of the models Firefox uses to sort tabs (as I had the same issues as you with the other model) and had amazing results. I learned quite a lot about how it all worked in the process, so if I could help in any way, please feel free to reach out.
> > 
> > https://github.com/Vertex-Mods/Zen-Tidy-Tabs
> 
> 
> 
> wow i just saw the screenshot and it looks good! i'll have a look at the code tomorrow, thank you!

I might suggest feeding it the url along with the title when you give it to the naming model, and if possible the tabs history, as Arc seems to do that too. I can also provide the prompts that Arc uses, I will drop the link here when I find it.

This is all the prompts TBC uses in there AI stuff: https://github.com/neurokitti/Arc-genAI

matteomekhail commented:
> > > @matteomekhail I'm not sure if it could be any help, but I vibecoded a project like this that used two of the models Firefox uses to sort tabs (as I had the same issues as you with the other model) and had amazing results. I learned quite a lot about how it all worked in the process, so if I could help in any way, please feel free to reach out.
> > > 
> > > https://github.com/Vertex-Mods/Zen-Tidy-Tabs
> > 
> > 
> > 
> > wow i just saw the screenshot and it looks good! i'll have a look at the code tomorrow, thank you!
> 
> I might suggest feeding it the url along with the title when you give it to the naming model, and if possible the tabs history, as Arc seems to do that too. I can also provide the prompts that Arc uses, I will drop the link here when I find it.
> 
> This is all the prompts TBC uses in there AI stuff: https://github.com/neurokitti/Arc-genAI
> 

Yep, I thought about that too, I also considered feeding the metadata from each tab’s meta tag, but that feels a bit over-engineered
The only issue I’ve found before with feeding the whole URL is that, on sites like YouTube, if a video has a title like “playing bla bla bla”, it treats it as if it were a playing website rather than a video player, and puts it in a different cluster.
I’ll try your solution tomorrow and see if I get the same behavior

matteomekhail committed:
feat: AI-powered tab grouping with local ML models

Uses Firefox's smart-tab-embedding and smart-tab-topic models to cluster tabs by semantic similarity and generate category labels based on URLs, titles, and browsing history.

matteomekhail commented:
I've implemented the solution using 12th suggestion, and it works much better now. It’s still not a perfect solution (ad example streaming instead of clothing, see video for reference) , but given the models available at the moment, I believe it’s the best possible solution, can always be upgraded later as smaller and better models will release and become available in Firefox.

Any suggestion to improve it is appreciate it! Tomorrow i'll make the UI look cleaner & add an animation while the model works

Attached is the video showing a demo

https://github.com/user-attachments/assets/277b9c81-e502-4e95-bc9a-0aa346ab0fb5

Thanks @12th-devs again :)

12th-devs commented:
Yeah of course, glad it worked out!

matteomekhail committed:
feat: enhance group tabs button UI with loading state and new icon

mr-cheffy commented:
Please let me know when its done for review!

matteomekhail commented:
> Please let me know when its done for review!

Sure it is ready now, Thank you!

mr-cheffy commented on src/browser/themes/shared/zen-icons/icons.css:
```diff
@@ -393,6 +393,18 @@
 }
 
+.zen-workspace-group-tabs-button {
+  list-style-image: none !important;
```
No need to add `!important` here?

mr-cheffy commented on src/zen/tabs/zen-tabs/vertical-tabs.css:
```diff
@@ -1364,3 +1364,49 @@
+/* ==========================================================================
+   Tab Category Headers (for Group function)
+   ========================================================================== */
+.tabbrowser-tab[zen-category-first="true"] {
```
This looks quite hacky to be honest, why not use something like firefox's existing tab group implementation?

matteomekhail replied:
<img width="304" height="699" alt="image" src="https://github.com/user-attachments/assets/cfd34885-4ef5-40c9-b64f-909759fe0c56" />

Well it would look like that and i'm not sure i like it? If you like it more i can leave it like that though

mr-cheffy replied:
We can change the styles of those. Just make the background always transparent and make it use the color we use for tab groups basically

12th-devs replied:
@matteomekhail if you need UI/UX for the tab groups, while it does not work anymore, an old project of mine [ATG](https://github.com/Vertex-Mods/Advanced-Tab-Groups) should help with both of those

matteomekhail replied:
<img width="299" height="619" alt="image" src="https://github.com/user-attachments/assets/c61522d7-8936-4f1a-b262-8abf6c58c4e0" />


I've just figure it out ahah,using the firefox group implementation it and changing the style it would look like that

mr-cheffy replied:
Looks good for now

mr-cheffy commented on src/zen/workspaces/ZenWorkspace.mjs:
```diff
@@ -14,6 +14,28 @@ class nsZenWorkspace extends MozXULElement {
+              <toolbarbutton command="cmd_zenGroupTabs"
+                             class="zen-workspace-group-tabs-button toolbarbutton-1">
+                <div class="zen-group-icon-container" style="display: grid; place-items: center; margin-inline-end: 5px;">
+                  <svg height="16px" width="16px" class="zen-group-loading-icon" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" style="grid-area: 1 / 1;">
```
Please don't use inline SVGs, rather use make a file and link it so it can be cached

matteomekhail replied:
Yep, done, also removed the spinner one as is not needed anymore

mr-cheffy commented on src/zen/workspaces/ZenWorkspaces.mjs:
```diff
@@ -2737,6 +2741,616 @@ class nsZenWorkspaces extends nsZenMultiWindowFeature {
     });
   }
 
+  async groupTabsBySimilarity() {
```
I'd move this into it's own `ZenTabsTidy.sys.mjs` file and just import it lazyly. maybe a new file under `zen/ml/ZenTabsTidy.sys.mjs`

matteomekhail replied:
Done, i moved the whole logic there

mr-cheffy commented on src/zen/workspaces/zen-workspaces.css:
```diff
@@ -6,6 +6,16 @@
 @namespace html 'http://www.w3.org/1999/xhtml';
 @namespace xul 'http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul';
 
+.zen-group-loading-icon {
```
Instead of a loading icon, shouldn't we make the separator a squiggly line going up and down? Like arc.

matteomekhail replied:
Done, i did the spinner before cause i though was harder to do, but it wasn't that hard after all

<img width="291" height="48" alt="image" src="https://github.com/user-attachments/assets/4aa6724f-8e66-4fa8-b4bc-ce04a05f83a4" />
<img width="232" height="38" alt="image" src="https://github.com/user-attachments/assets/4d792c1e-6391-4a54-a3e6-e74f990ce356" />

mr-cheffy commented on src/zen/workspaces/ZenWorkspace.mjs:
```diff
@@ -1,10 +1,10 @@
-class nsZenWorkspace extends MozXULElement {
-  static get markup() {
-    return `
+{
```
Also, seems like something has gone wrong while migrating the branch?

matteomekhail replied:
i think it had something to do when i fix the conflicts, whenever you review the new changes i'll check again to merge it

matteomekhail committed:
Removed unnecessary !important

matteomekhail committed:
Used nowfirefox's existing tab group implementation instead of custom category implementation

matteomekhail committed:
Icons separeted and not SVG inline now

matteomekhail committed:
Fix: Moved the ML Logic to an extra file

matteomekhail committed:
feat: enhance vertical tabs UI with new separator styles and animations for grouping, now like arc has the squiggly line, removed the spinner

mr-cheffy commented on src/zen/tabs/zen-tabs/vertical-tabs.css:
```diff
@@ -146,12 +146,36 @@
       border-inline-start: none;
     }
+
+    &::after {
```
I'd only define this element once data-grouping is enabled

matteomekhail replied:
Yep good idea, done

mr-cheffy commented on src/zen/ml/ZenTabsTidy.sys.mjs:
```diff
@@ -0,0 +1,647 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+const Services = globalThis.Services;
```
`Services` will always be global, we don't need this

matteomekhail replied:
Removed now, thanks

mr-cheffy commented on src/zen/ml/ZenTabsTidy.sys.mjs:
```diff
@@ -0,0 +1,647 @@
+
+  try {
+    const { createEngine } = ChromeUtils.importESModule(
+      'chrome://global/content/ml/EngineProcess.sys.mjs'
```
Maybe lazyly load this module?

https://github.com/zen-browser/desktop/blob/dev/src/zen/urlbar/ZenUBActionsProvider.sys.mjs#L23-L30

matteomekhail replied:
Not sure i've done it correctly, please double check, but should work

mr-cheffy commented on src/zen/ml/ZenTabsTidy.sys.mjs:
```diff
@@ -0,0 +1,647 @@
+    const keywords = extractKeywords(titles);
+    const domains = [...new Set(clusterData.map((data) => data.domain))].join(', ');
+
+    const prompt = `You are an expert organizer who creates concise, descriptive category names.
```
Not sure if this is the best way we should do it, specially since there might be some possible vulnerability by directly injecting the scripts. Firefox's implementation (https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs) doesn't do this, for example

matteomekhail replied:
I don't see any possibility to inject scripts here?
also i believe that firefox implementation is very similar,  we both import createEngine, call the model, even tho it uses a much smaller prompt

can you please tell me what worries you exactly so i can fix it? or if you have some alternatives in mind?

mr-cheffy replied:
> can you please tell me what worries you exactly so i can fix it? or if you have some alternatives in mind?
Im still investigating how firefox does it as they dont have any sort of "huge text" chunk going on

matteomekhail replied:
> I think it's because Mozilla Firefox uses this https://huggingface.co/Mozilla/smart-tab-topic for labeling

i am also using this?

<img width="541" height="306" alt="image" src="https://github.com/user-attachments/assets/5d74283c-69a4-4d57-bc90-a1d53e9d2b26" />
<img width="444" height="147" alt="image" src="https://github.com/user-attachments/assets/ec5ca534-e53c-4bec-a7ca-e26dd7ac5d33" />


I'm actually using both smart-tab-embedding & smart-tab-topic, one to see how similar the tabs are, one to generate the label, not sure if is the best option thought

Minemetero replied:
Firefox’s SmartTabGrouping https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs#1199-1210 only sends a tiny template to the model and then cleans up the result with a bunch of JS
ZenTabsTidy just puts all those rules directly into the prompt instead
So they basically achieve the same thing

matteomekhail replied:
Yep, correct

> Firefox’s SmartTabGrouping https://searchfox.org/firefox-main/source/browser/components/tabbrowser/SmartTabGrouping.sys.mjs#1199-1210 only sends a tiny template to the model and then cleans up the result with a bunch of JS ZenTabsTidy just puts all those rules directly into the prompt instead So they basically achieve the same thing

mr-cheffy reviewed (changes requested):
Good, but instead of adding the new file to the jar manifest, better just make a new moz.build on that directory and declare it as a system module.

For example: 
https://github.com/zen-browser/desktop/blob/886b559b507438c38b640dce0657bd627acfeb51/src/zen/common/moz.build#L6

matteomekhail committed:
refactor: update vertical tabs CSS by removing deprecated styles and enhancing separator animations for grouping

matteomekhail committed:
refactor: remove unused Services import from ZenTabsTidy.sys.mjs

matteomekhail committed:
refactor: optimize EngineProcess import by using lazy loading in ZenTabsTidy.sys.mjs

matteomekhail committed:
refactor: update ZenTabsTidy import path and clean up zen-assets.jar.inc.mn

12th-devs commented:
What all still needs to happen here?

mr-cheffy committed:
Discard changes to locales/en-GB/browser/browser/zen-workspaces.ftl

mr-cheffy committed:
Discard changes to locales/it/browser/browser/zen-workspaces.ftl

mr-cheffy commented on src/browser/themes/shared/zen-icons/icons.css:
```diff
@@ -393,6 +393,18 @@
+    display: none !important;
+  }
+
+  @media -moz-pref('zen.view.show-clear-tabs-button', false) {
```
Should also be hidden if `browser.ml.enable` is disabled

matteomekhail replied:
Yes agree, i just fixed that

matteomekhail committed:
fix: hide group tabs button when browser.ml.enable is disabled

12th-devs commented:
@matteomekhail from what I remember, the AI model sorted the tabs really fast, so I added a minimum time in the UI as it made it feel more natural, because of it doing the sorting instantly and seemed unsettling to me almost. Just something to consider based on your testing

matteomekhail commented:
> @matteomekhail from what I remember, the AI model sorted the tabs really fast, so I added a minimum time in the UI as it made it feel more natural, because of it doing the sorting instantly and seemed unsettling to me almost. Just something to consider based on your testing

That's probably a good idea, i don't want it to feel fake thought, i will try to see what it feels better

matteomekhail committed:
fix: increase timeout duration for tab grouping completion in ZenTabsTidy

matteomekhail commented:
> That's probably a good idea, i don't want it to feel fake thought, i will try to see what it feels better

Increased it from 500 to 750ms, it feels a bit better now, 1s was way too much based on my testing, agreed that 500ms was maybe too fast

peteygao commented:
It only feels fast because you're testing with a small number of tabs. Increase the tab count to something reasonable (100+ for starters) where you'd needing a grouping feature and it'll start to crawl (transformer models slows down quadratically as context lengths increases). I wouldn't add any artificial delay to the grouping functionality.

12th-devs commented:
> It only feels fast because you're testing with a small number of tabs. Increase the tab count to something reasonable (100+ for starters) where you'd needing a grouping feature and it'll start to crawl (transformer models slows down quadratically as context lengths increases). I wouldn't add any artificial delay to the grouping functionality.

I agree with the context length stuff, that's a known issue with these models, but respectfully, I disagree with you about the artificial delay. If you sort 100 tabs and it takes 5 seconds, the delay is gonna be ignored, but if it's 10 tabs and it's sorted in 0.3 seconds, then the 1 second delay is gonna make it feel a bit more real. Does that clear things up a bit?

mr-cheffy commented:
> then the 1 second delay is gonna make it feel a bit more real. Does that clear things up a bit?

Users don't care if it feels real or not, artificial delay means users will always be waiting for X amounts of seconds and that can become tiresome in the long run

BlueFox1616 commented:
Leaving this here because i thing Arc has a potentially better method for gathering more info about each tab 
https://github.com/neurokitti/Arc-genAI
I think Arc also utilizes the tabs history as far as i remember

jacksongoode commented:
> Leaving this here because i thing Arc has a potentially better method for gathering more info about each tab [neurokitti/Arc-genAI](https://github.com/neurokitti/Arc-genAI?rgh-link-date=2026-01-04T00%3A43%3A57Z) I think Arc also utilizes the tabs history as far as i remember

I don't think this needs to be included in this PR, but Arc/Dia in addition does a few nice things that could eventually be done in this implementation. This list was a result of an LLM review - not directly by me. Still worth adding to this thread given the feature could benefit from it:

   * Similarity consideration into the algorithm when two tabs share the same opener (parent-child relationship).
   * Factor the time tabs were opened into the clustering logic; tabs opened in a quick sequence are often part of the same mental task.
   * Prioritize the first path segment of a URL (e.g., github.com/zen-browser/) over the page title.
   * The local topic-generation prompt could look for a specific "Task" or "Project" name rather than just the most frequent keyword - considering intent.
   * Consider the "active tab" at the time of opening a new tab as a strong semantic anchor for the new cluster.
